Prepare 10.1 release (#633)
rescript-association/rescript-lang.org: _blogposts/2023-02-02-release-10-1.mdx

---
author: rescript-team
date: "2023-02-02"
title: ReScript 10.1
badge: release
description: |
  Async/await & better Promise support, JSX v4, and more!
---

## Introduction

We are happy to announce ReScript 10.1!

ReScript is a robustly typed language that compiles to efficient and human-readable JavaScript. It comes with one of the fastest build toolchains and offers first class support for interoperating with ReactJS and other existing JavaScript code.

> Use `npm` to install the newest [10.1 release](https://www.npmjs.com/package/bs-platform/v/10.1.1):

Reviewer: rename bs-platform -> rescript

Author: fixed

@val external fetchUserMail: string => promise<string> = "GlobalAPI.fetchUserMail"
@val external sendAnalytics: string => promise<unit> = "GlobalAPI.sendAnalytics"

// We use the `async` keyword to allow the use of `await` in the function body
let logUserDetails = async (userId: string) => {
  // We use `await` to fetch the user email from our fictive user endpoint
  let email = await fetchUserMail(userId)

  await sendAnalytics(`User details have been logged for ${userId}`)

  Js.log(`Email address for user ${userId}: ${email}`)
}
```

To learn more about our async / await feature, check out the relevant [manual section](/docs/manual/latest/async-await).

## New `promise` builtin type and `Js.Promise2` module

In previous versions of ReScript, promises were expressed as a `Js.Promise.t<'a>` type, which was a little tedious to type. From now on, users may use the `promise<'a>` type instead.

Quick example of a `.resi` file using the new `promise` type:

```resi
// User.resi
type user

let fetchUser: string => promise<user>
```

Way easier on the eyes, don't you think? Note that the new `promise` type is fully compatible with `Js.Promise.t` (no breaking changes).

> Additionally, we also introduced the `Js.Promise2` module as a stepping stone to migrate `Js.Promise` based code to a first-pipe (->) friendly solution.

Author: Yes we could do that. Will think about it.

> Additionally, we also introduced the `Js.Promise2` module as a stepping stone to migrate `Js.Promise` based code to a first-pipe (->) friendly solution.

Author: done

(*Sidenote*: We know that our users want a solution to unify `Belt`, `Js` and `Js.xxx2` and have a fully featured "standard library". Good news is that we have a solution in the pipeline to fix this. `Js.Promise2` was introduced to ease the process later on and is not supposed to be the panacea of promise handling.)

If you are already using a third-party promise library like [ryyppy/rescript-promise](https://github.com/ryyppy/rescript-promise) or similar, there's no need to migrate, but we highly encourage you to mix and match it with `async` / `await`.

## New JSX v4 syntax

> ReScript 10.1 now ships with JSX v4. Here's what's new:

Author: Good point. Will add.

> ReScript 10.1 now ships with JSX v4. Here's what's new:

Author: done.

- **Cleaner interop.** Due to recent improvements in the type checker, the `@react.component` transformation doesn't require any `makeProps` convention anymore. `make` functions will now be transformed into a `prop` type and a component function. That's it.
- **Two new transformation modes**. JSX v4 comes with a `classic` mode (= `React.createElement`) and `automatic` mode (= `jsx-runtime` calls). The latter is the new default, moving forward with `rescript/react@0.11` and `React@18`.
- **Allow mixing JSX configurations on the project and module level.** Gradually mix and match JSX transformations and modes without migrating any old code!
- **Pass `prop` types** to `@react.component`. You can now fine tune `@react.component` with your specific prop type needs. Very useful for libraries and frameworks to define component interfaces.
- **Revisited props spread operator.** This will allow users to spread records in JSX without sacrificing their sanity. Note that this implementation has harder constraints than its JS counterpart. (requires `rescript/react@0.11` or higher)

> 

Reviewer 2: ```suggestion
- **Paving the way for better editor tooling.** The new representation in JSX v4 opens up for better editor tooling, like providing typed completion for values and callbacks of JSX props.
- **Better type inference of props.** Type inference when passing for example variants that are defined in the same module as the component is much improved. With the earlier JSX version, you'd often need to write code like this in order for the compiler to understand which variant you're passing: `<Button variant=Button.Primary text="Click" />`. With JSX v4, you won't need to tell the compiler where the variant you're passing is located: `<Button variant=Primary text="Click" />`

```

Code tells more than words, so here's a non-exhaustive code example to highlight the different JSX features. Make sure to also check out the JS output and play around with the code in our newest playground!

<CodeTab labels={["ReScript", "JS Output"]}>

```res
// Set the jsx configuration per module
@@jsxConfig({version: 4, mode: "automatic"})

module AutomaticModeExample = {
  // "automatic" mode will compile jsx to the React 18 compatible
  // jsx-runtime calls
  @@jsxConfig({version: 4, mode: "automatic"})

  @react.component
  let make = (~name) => {
    <div> {React.string(`Hello ${name}`)} </div>
  }
}

module ClassicModeExample = {
  // "classic" mode will compile jsx to React.createElement calls
  @@jsxConfig({version: 4, mode: "classic"})

  @react.component
  let make = (~name) => {
    <div> {React.string(`Hello ${name}`)} </div>
  }
}

> module NoDecoratorExample = {

Author: We can rename it.

> module NoDecoratorExample = {

Author: done.

  }
}

module ReactInterfaceExample: {
  @react.component
  let make: (~name: string, ~age: int=?) => React.element
} = {
  @react.component
  let make = (~name, ~age=0) => {
    <div>
      {React.string(
        `Hello ${name}, you are ${Belt.Int.toString(age)} years old.`,
      )}
    </div>
  }
}

module PropTypeInjectionExample = {
  // Let's assume we have a prop type that we wanna enforce
  // as our labeled arguments
  type someoneElsesProps = {isHuman: bool}

  // Here we tell the `react.component` decorator what props to infer.
  // Useful for e.g. NextJS usage, or to create components that should
  // comply to certain library component interfaces
  @react.component(: someoneElsesProps)
  let make = (~isHuman) => {
    let msg = switch isHuman {
    | true => "hello human"
    | false => "hello fellow computer"
    }
    <div> {React.string(msg)} </div>
  }
}

module PropSpreadExample = {
  // Note: This will require @rescript/react 0.11 or later
  @@jsxConfig({version: 4, mode: "automatic"})

  @react.component
  let make = () => {
    let props = {NoDecoratorExample.name: "World"}

    <NoDecoratorExample {...props} />
  }
}

> 

Author: Add another example for better type inference on nominal types:
https://rescript-lang.org/try?version=v10.1.2&code=ALBWGcA8GEHsDsBmBLA5gCgN4AIBuBTAJ3GQQC5sBmAGmwFtYATfCgIgEMBXAF1jve7IAxq2wBfAJQAoKQ0acANvmwBVcEWwBebJinZs3AJ4AHZeG4DO4LdgDy8BcnjKAPncSJHzvdh-BC+OxC3AB0QnzGCPjw3D5K3PTsANbK2ugAfuaW4BJaAHw6PvrgAO7I3EIAFthZ3FaF+o3YbvZeqQUASoHBIfCKCkWNLR5t+dhdQaF9CgNNYj7z81Lx2AD6NgA8ahq1VpqtTsoA9HlSQA

Author: no need anymore, the improvement has been mentioned above already.

            });
}

var PropSpreadExample = {
  make: Playground$PropSpreadExample
};

var root = JsxRuntime.jsxs("div", {
      children: [
        JsxRuntime.jsx(Playground$AutomaticModeExample, {
              name: "Automatic"
            }),
        JsxRuntime.jsx(Playground$ClassicModeExample, {
              name: "Classic"
            }),
        JsxRuntime.jsx(make, {
              name: "NoDecorator"
            }),
        JsxRuntime.jsx(Playground$ReactInterfaceExample, {
              name: "Interface"
            }),
        JsxRuntime.jsx(Playground$PropTypeInjectionExample, {
              isHuman: true
            }),
        JsxRuntime.jsx(Playground$PropSpreadExample, {})
      ]
    });

export {
  AutomaticModeExample ,
  ClassicModeExample ,
  NoDecoratorExample ,
  ReactInterfaceExample ,
  PropTypeInjectionExample ,
  PropSpreadExample ,
  root ,
}
```
</CodeTab>

### How to migrate to JSX v4?

We provide a full [migration guide](/docs/react/latest/migrate-react) with all the details of an migration.

Make sure to also check out the [rescript-react changelog](https://github.com/rescript-lang/rescript-react/blob/master/CHANGELOG.md) as well.

> 

Author: Add acknowledgements for the JSX@v4 work

Author: done.

## What's next?

> Our contributors are already one step ahead and work on major improvements in our next upcoming major release (v11). Things that are currently being explored:

Author: Eh, sounds weird.. maybe...

```suggestion
Our contributors are already one step ahead and are currently working on improvements for the next major v11 release. Things that are currently being explored:
```
